pcdsdevices/ccm.py

```diff
@@ -773,6 +773,178 @@ def inverse(self, real_pos: namedtuple) -> namedtuple:
+    ):
+        """
+        Simple monotonic search for the CCM energy position:
+        Move CCM slightly under the setpoint and check if the downstream/upstream IPM
+        average is above a given threshold. If not, move the CCM energy up a bit and check
+        again.
+
+        Parameters
+        ----------
+        setpoint : float
+            Energy setpoint
+        offset: float
+            Energy offset. The initial CCM energy will be setpoint-offset.
+        n_average: int
+            Number of data points in the IPM average.
+        threshold_ratio: float
+            If mean(IPM downstream)/mean(IPM upstream)>threshold_ratio, the optimization
+            finishes.
+        stepsize: float
+            Energy increment after a failed check on the IPM ratio. Should be smaller
+            than the self-seed bandwidth for good results.
+        max_steps: int
+            If no good point is reached after max_step points, the optimization is
+            aborted.
+        """
+        self.self_seed.move(setpoint)
+        # move slightly below the setpoint
+        super().move(setpoint-offset, wait=True)
+        # wait for self-seed to settle
+        time.sleep(1)
+
+        for step in range(max_steps):
+            intensities = self.get_ipms_intensities(n_average)
+            ipm_ratio = self.ipm_ratio(intensities)
+            if ipm_ratio > self.ipm_ratio_threshold:
+                print('Optimization complete!\n')
+                break
+            super().mvr(stepsize, wait=True)
+        return
+
+    def get_ipms_intensities(self, nn, sleep=0.01):
+        intensities = []
+        for ii in range(nn):
+            intensities.append(self.ipms.get())
+            time.sleep(sleep)
+        return np.asarray(intensities)
+
+    def ipm_ratio(self, intensities):
+        """ Calculate the mean(downstream)/mean(upstream) IPMs """
+        ratio = np.average(intensities, axis=0)
+        ratio = ratio[self.ipm_idx['downstream']] / ratio[self.ipm_idx['upstream']]
+        return ratio
+
+    def setup_method_monotonic(self):
+        print('Setting up monotonic search parameters.')
+
+        s = 'Offset (eV) (default: TBD): '
+        offset = float(input(s) or 5)
+
+        s = 'Number of shots for the IPMs average (default: 50): '
+        n = int(input(s) or 50)
+
+        s = 'IPM ratio threshold (default: TBD): '
+        ratio_th = float(input(s) or 2.5)
+
+        s = 'Energy increment (eV) (default: TBD): '
+        step_size = float(input(s) or 0.5)
+
+        s = 'Maximum number of optimization steps (default: 10): '
+        max_step = int(input(s) or 10)
+
+        self.optimization_args = {
+            'energy_offset' : offset,
+            'n_average' : n,
+            'ipm_ratio_threshold' : ratio_th,
+            'energy_step_size' : step_size,
+            'max_step' : max_step
+        }
+        self.method = 'monotonic'
+        return
+
+    def move(self, new_pos , *args, **kwargs):
```

Comment:
For this positioner to work with bluesky plans, you'll likely need to hijack the `.set()` method as well.  That's the method that plans like `bp.scan` will use, I believe

Reply:
See Zach's comment for solution